**Code reviewer**

¡Hola en esta nueva aventura!

Mi nombre es **Ivan Hernandez** y seré tu revisor de código para este proyecto.

Revisaré todo el cuaderno con preguntas o comentarios dejados atrás. Te animo a responder los comentarios con cualquier duda o solución a la recomendación o problema presentado **(con la celda en color azul)**, esa es la mejor manera de entender y aprender en este nuevo mundo de datos y programación. Seguiré la misma lista de verificación y los mismos pasos en los criterios de evaluación del proyecto.

 Importante, <span style="color:red">por favor no cambies mis comentarios</span>, es una buena forma de seguir tu aprendizaje por mí y otros revisores de código y finalmente, pongo un ejemplo del color del código y su significado.

<div class="alert alert-success">
<h2> Comentarios del revisor </h2>
Excelente trabajo y ejecución
</div>

<div class="alert alert-warning">
<h2> Comentarios del revisor </h2>
Comentario no crítico; si hay menos de tres de tales comentarios, el proyecto puede aceptarse sin su refinamiento.</div>

<div class="alert alert-danger">
<h2> Comentarios del revisor </h2>
Error que debe corregirse, de lo contrario el proyecto no será aceptado.
</div>

<div class="alert alert-info">
<h2> Comentarios del estudiante</h2>
Comentario de ejemplo
</div>

**Retroalimentación general  V1**

Hay un par de errores importantes a corregir que dejé descritos, espero que sirvan de mucha ayuda y si estás muy perdido no dudes en escribirle a tú tutor/instructor con el pantallazo de los errores descritos en el proyecto. 

Mucha suerte!

**Retroalimentación general  V2**

Muy buen proyecto, todo claro y las correcciones hechas.

# Análisis del riesgo de incumplimiento de los prestatarios

Tu proyecto consiste en preparar un informe para la división de préstamos de un banco. Deberás averiguar si el estado civil y el número de hijos de un cliente tienen un impacto en el incumplimiento de pago de un préstamo. El banco ya tiene algunos datos sobre la solvencia crediticia de los clientes.

Tu informe se tendrá en cuenta al crear una **puntuación de crédito** para un cliente potencial. La **puntuación de crédito** se utiliza para evaluar la capacidad de un prestatario potencial para pagar su préstamo.

En el siguiente proyecto intentare considerar si los clientes son aptos o elegibles para un prestamo, dependiendo de diferentes caracteristicas que puedan alterar su eligibilidad.

<div class="alert alert-danger">
<h2> Comentarios del revisor </h2>

Justo como dice el enunciado, no olvides eliminar los comentarios con corchetes cuadrados [].
</div>

<div class="alert alert-info">
<h2> Comentarios del estudiante</h2>
Hecho.
</div>

## Abre el archivo de datos y mira la información general. 

Comienzo por importar las librerias y leer el siguiente csv.

In [1]:
import pandas as pd

In [2]:
general_data = pd.read_csv('/datasets/credit_scoring_eng.csv')

<div class="alert alert-danger">
<h2> Comentarios del revisor </h2>

Importante que separes la importación de la carga de datos en celdas diferentes, luego cuando trabajes con muchas librerías te ayudará a tener más orden.
</div>

<div class="alert alert-info">
<h2> Comentarios del estudiante</h2>
Listo.
</div>

## Ejercicio 1. Exploración de datos

**Descripción de los datos**
- `children` - el número de hijos en la familia
- `days_employed` - experiencia laboral en días
- `dob_years` - la edad del cliente en años
- `education` - la educación del cliente
- `education_id` - identificador de educación
- `family_status` - estado civil
- `family_status_id` - identificador de estado civil
- `gender` - género del cliente
- `income_type` - tipo de empleo
- `debt` - ¿había alguna deuda en el pago de un préstamo?
- `total_income` - ingreso mensual
- `purpose` - el propósito de obtener un préstamo

Se puede notar que hay 21525 entradas, es decir, filas, y 12 columnas.

In [3]:
general_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
general_data.head(10)



children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

Se nota que hay valores ausentes al ver que hay ciertos datos que no se muestran o no son simetricos con las demas columnas, en este caso en days_employed y total_income.

In [5]:
general_data.isnull().sum()


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Parece que los valores ausentes son simetricos, pues tienen algun tipo de correlacion ya que siempre cuando days_employed es NaN, total_income tambien lo sera.

In [6]:
general_data[general_data['days_employed'].isna()]



children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

Pues parece que la hipotesis de que ambas columnas tienen una correlacion, por lo que se puede hacer mayor investigacion para poder ver si tienen algun tipo de relacion con otro tipo de columnas, de esta manera podiendo ayudar en el proceso de lidiar con los valores ausentes.

In [7]:
missing_values = general_data[(general_data['days_employed'].isna()) & (general_data['total_income'].isna())]
missing_values.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          2174 non-null   int64  
 1   days_employed     0 non-null      float64
 2   dob_years         2174 non-null   int64  
 3   education         2174 non-null   object 
 4   education_id      2174 non-null   int64  
 5   family_status     2174 non-null   object 
 6   family_status_id  2174 non-null   int64  
 7   gender            2174 non-null   object 
 8   income_type       2174 non-null   object 
 9   debt              2174 non-null   int64  
 10  total_income      0 non-null      float64
 11  purpose           2174 non-null   object 
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB


<div class="alert alert-success">
<h2> Comentarios del revisor </h2>

Muy buen trabajo!
</div>

**Conclusión intermedia**

Pues para poder ver la relacion es mejor sacar un porcentaje de la cantidad de valores ausentes en relacion a la totalidad de datos en el dataframe, por lo que voy a usar f-strings para poder mostrar el porcentaje de valores ausentes. Si este llegase a ser mas de 10%, puede significar que tenga algo de relevancia para nuestra investigacion.

En este caso voy a tomar de referencia a la columna de "education", pues creo que puede tener algo que ver con el tipo de educacion que tiene cada individuo.

In [8]:
total_missing_values = (len(general_data[general_data['days_employed'].isna()])) & (len(general_data[general_data['total_income'].isna()]))
total_values = len(general_data)
missing_values_percentage = (total_missing_values/total_values)
print(f'El porcentaje de valores ausentes en comparacion a la totalidad de datos es: {missing_values_percentage:%}.')

El porcentaje de valores ausentes en comparacion a la totalidad de datos es: 10.099884%.


In [9]:
general_data['education'].unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

In [10]:
general_data['education'] = general_data['education'].str.lower()
general_data['education'].unique()

array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

In [11]:
second_edu_mv = len(general_data[(general_data['education'] == "secondary education") & (general_data['days_employed'].isna())])
prim_edu_mv = len(general_data[(general_data['education'] == "primary education") & (general_data['days_employed'].isna())])
bach_degree_mv = len(general_data[(general_data['education'] == "bachelor's degree") & (general_data['days_employed'].isna())])
some_college_mv = len(general_data[(general_data['education'] == "some college") & (general_data['days_employed'].isna())])
total_ed_missing_values = (second_edu_mv+prim_edu_mv+bach_degree_mv+some_college_mv)
print(second_edu_mv)
print(prim_edu_mv)
print(bach_degree_mv)
print(some_college_mv)
total_ed_missing_values

1540
21
544
69


2174

In [12]:
general_data[(general_data['education'] == "graduate degree")]
# Vamos a investigar a los clientes que no tienen datos sobre la característica identificada y la columna con los valores ausentes



children  days_employed  dob_years        education  education_id  \
2963          0  337584.815560         69  graduate degree             4   
4170          0    -409.200149         45  graduate degree             4   
6551          0   -5352.038180         58  graduate degree             4   
12021         3   -5968.075884         36  graduate degree             4   
12786         0  376276.219531         62  graduate degree             4   
21519         1   -2351.431934         37  graduate degree             4   

      family_status  family_status_id gender    income_type  debt  \
2963        married                 0      M        retiree     0   
4170      unmarried                 4      M       employee     0   
6551        married                 0      M       employee     0   
12021       married                 0      F  civil servant     0   
12786       married                 0      F        retiree     0   
21519      divorced                 3      M       employee     0   

       total_income                                   purpose  
2963      15800.399               buy residential real estate  
4170      31771.321  transactions with commercial real estate  
6551      42945.794                       going to university  
12021     17822.757                     purchase of the house  
12786     40868.031               buy residential real estate  
21519     18551.846                buy commercial real estate

In [13]:
# Comprobación de la distribución
pivot_table1= general_data.pivot_table(values= general_data[(general_data['days_employed'].isna()) & (general_data['total_income'].isna())], index= 'education', aggfunc= 'count')
pivot_table1

children  days_employed   debt  dob_years  education_id  \
education                                                                      
bachelor's degree        5260           4716   5260       5260          5260   
graduate degree             6              6      6          6             6   
primary education         282            261    282        282           282   
secondary education     15233          13693  15233      15233         15233   
some college              744            675    744        744           744   

                     family_status  family_status_id  gender  income_type  \
education                                                                   
bachelor's degree             5260              5260    5260         5260   
graduate degree                  6                 6       6            6   
primary education              282               282     282          282   
secondary education          15233             15233   15233        15233   
some college                   744               744     744          744   

                     purpose  total_income  
education                                   
bachelor's degree       5260          4716  
graduate degree            6             6  
primary education        282           261  
secondary education    15233         13693  
some college             744           675

Teniendo en cuenta que el porcentaje de "graduate degree" referente al total de valores es muy bajo, puede significar que haya una correlación alta si tomamos otros valores que no sean "graduate degree", pero para ser mas  certeros se puede seguir buscando patrones.]

**Posibles razones por las que hay valores ausentes en los datos**


Puedo decir que hasta ahora parece que los valores ausentes parecen ser aleatorios, pues gran relacion entre los datos de "education" no hay. Se puede seguir buscando patrones, pero probablemente sean un error aleatorio.


**Conclusión intermedia**

Voy a buscar patrones en el tipo de trabajo que tiene cada cliente para ver si tiene relevancia con los valores ausentes, pues son los ingresos y dias de trabajo, cosa que puede tener relacion con el empleo de cada individuo.

In [14]:
# Comprueba otras razones y patrones que podrían llevar a valores ausentes
employee_mv = len(general_data[(general_data['income_type'] == "employee") & (general_data['days_employed'].isna())])
retiree_mv = len(general_data[(general_data['income_type'] == "retiree") & (general_data['days_employed'].isna())])
business_mv = len(general_data[(general_data['income_type'] == "business") & (general_data['days_employed'].isna())])
civil_servant_mv = len(general_data[(general_data['income_type'] == "civil servant") & (general_data['days_employed'].isna())])
unemployed_mv = len(general_data[(general_data['income_type'] == "unemployed") & (general_data['days_employed'].isna())])
entrepreneur_mv = len(general_data[(general_data['income_type'] == "entrepreneur") & (general_data['days_employed'].isna())])
student_mv = len(general_data[(general_data['income_type'] == "student") & (general_data['days_employed'].isna())])
paternity_maternity_mv = len(general_data[(general_data['income_type'] == "paternity / maternity leave") & (general_data['days_employed'].isna())])
print(employee_mv)
print(retiree_mv)
print(business_mv)
print(civil_servant_mv)
print(unemployed_mv)
print(entrepreneur_mv)
print(student_mv)
print(paternity_maternity_mv)

total_income_type_missing_values = (employee_mv+retiree_mv+business_mv+civil_servant_mv+unemployed_mv+entrepreneur_mv+student_mv+paternity_maternity_mv)
print(f'El porcentaje de relacion entre "income_type" y valores ausentes es: {(total_income_type_missing_values/total_values):%}')

1105
413
508
147
0
1
0
0
El porcentaje de relacion entre "income_type" y valores ausentes es: 10.099884%


**Conclusión intermedia**

Se puede confirmar que los valores ausentes son accidentales, pues la relevancia entre el tipo de educacion y tipo de empleo dicen que no tiene mucha relevancia estos datos con los valores ausentes.

In [15]:
# Comprobación de otros patrones: explica cuáles
general_data[(general_data['income_type'] == "employee") & (general_data['days_employed'].isna())]

children  days_employed  dob_years            education  education_id  \
82            2            NaN         50    bachelor's degree             0   
83            0            NaN         52  secondary education             1   
90            2            NaN         35    bachelor's degree             0   
96            0            NaN         44  secondary education             1   
97            0            NaN         47    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
21432         1            NaN         38         some college             2   
21463         1            NaN         35    bachelor's degree             0   
21495         1            NaN         50  secondary education             1   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
82               married                 0      F    employee     0   
83               married                 0      M    employee     0   
90               married                 0      F    employee     0   
96               married                 0      F    employee     0   
97               married                 0      F    employee     0   
...                  ...               ...    ...         ...   ...   
21432          unmarried                 4      F    employee     0   
21463  civil partnership                 1      M    employee     0   
21495  civil partnership                 1      F    employee     0   
21502            married                 0      F    employee     0   
21510            married                 0      F    employee     0   

       total_income                      purpose  
82              NaN                      housing  
83              NaN                      housing  
90              NaN         housing transactions  
96              NaN  buy residential real estate  
97              NaN            profile education  
...             ...                          ...  
21432           NaN         housing transactions  
21463           NaN             having a wedding  
21495           NaN             wedding ceremony  
21502           NaN       building a real estate  
21510           NaN                 car purchase  

[1105 rows x 12 columns]

**Conclusiones**


Se encontro que cuando son trabajadores (employee), o tienen hasta educacion secundaria tienen mas valores ausentes. Con esto logro ver que no hay un patron como tal, pero es util indagar mas datos sobre los valores ausentes.

Los valores ausentes se pueden lidiar de varias maneras, en este caso intentare agrupar datos que tengan relevancia con "total_income" y "days_employed". Al terminar de agruparlos se podra lograr reemplazar los datos con mediana o promedio, dependiendo cual convenga mas.



## Transformación de datos


# Los siguientes pasos de educacion ya se hicieron previamente

In [16]:
# Veamos todos los valores en la columna de educación para verificar si será necesario corregir la ortografía y qué habrá que corregir exactamente


In [17]:
# Arregla los registros si es necesario


In [18]:
# Comprobar todos los valores en la columna para asegurarnos de que los hayamos corregido



In [19]:
general_data['children'].unique()


array([ 1,  0,  3,  2, -1,  4, 20,  5])

Hay datos que son erroneos o muy poco probables, como tener negativo numero de hijos, o tener 20 hijos. Para reemplazar estos datos, voy a reemplazar estos valores con la mediana de "children" ya que un promedio daria un dato float, y no se puede tener un hijo y medio.

In [20]:
children_median = general_data['children'].median()
children_median_int = children_median.astype('int64')
general_data['children'] = general_data['children'].replace([20, -1], children_median_int)
# [arregla los datos según tu decisión]


In [21]:
general_data['children'].unique()
# Comprobar la columna `children` de nuevo para asegurarnos de que todo está arreglado



array([1, 0, 3, 2, 4, 5])

<div class="alert alert-success">
<h2> Comentarios del revisor </h2>

Muy buen trabajo!
</div>

[Hay datos negativos en "days employed" lo cual es imposible, pues puedes o haber trabajado en tu vida o no.]

# Encuentra datos problemáticos en `days_employed`, si existen, y calcula el porcentaje


Se puede visualizar que los dias de empleo son valores muy altos, por lo que se tomara en cuenta que estan escritos en horas, entonces se dividira aquellos que superen 60 años que es lo minimo que puede haber trabajado uno de los clientes, pues tiene 75 años de edad (asumiendo que se trabaja desde los 15), por 24, lo que es un día en horas

In [22]:
general_data['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

In [23]:
negative_days_employed = len(general_data[general_data['days_employed'] < 0])
print(f'El porcentaje de valores erroneos en days employed es {(negative_days_employed/total_values):%}.')

El porcentaje de valores erroneos en days employed es 73.895470%.


In [24]:
# Aborda los valores problemáticos, si existen.
general_data['days_employed'] = general_data['days_employed'].abs()

In [25]:
def hours_to_day_days_employed(x):
    if x > 2195:
        return (x/24)
    else:
        return x

general_data['days_employed'] = general_data['days_employed'].apply(hours_to_day_days_employed)

In [26]:
general_data['days_employed'].describe()

count    19351.000000
mean      3258.411829
std       5600.674988
min         24.141633
25%        192.373997
50%        660.253676
75%       1723.578811
max      16739.808353
Name: days_employed, dtype: float64

In [27]:
# Comprueba el resultado - asegúrate de que esté arreglado
general_data[general_data['days_employed'] < 0].count()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


Aparecen datos que contienen edad '0', y pues no pueden haber personas con edad 0 en un conjunto de datos, ya que serian no existentes. Para solucionarlo se usara la mediana, pues se esta usando las edades como integros, y el promedio daria un dato tipo float.

<div class="alert alert-danger">
<h2> Comentarios del revisor V1</h2>

Una vez que aplicaste describe, observas que hay valores anormalmente altos, que corresponden a 1100 años o 401755 días. En este caso debes imputarlos o asumir que están en horas y corregirlos. Lo puedes hacer a través del método `.loc[condicional, 'days_employed']` y asignar el valor correcto. 
    
Es muy importante corregirlo, debido a que estás usando estos datos para generar conclusiones.
</div>

<div class="alert alert-info">
<h2> Comentarios del estudiante</h2>
Listo, transformados a dias considerando que estan en horas y que el maximo de edad es 75, pues se considero como 60 años como la cantidad máxima de años de trabajo.
</div>

In [28]:
general_data['dob_years'].unique()

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51,  0, 59, 29, 60, 55, 58, 71, 22, 73,
       66, 69, 19, 72, 70, 74, 75])

In [29]:
dob_years_median = general_data['dob_years'].median()
dob_years_median_int = dob_years_median.astype('int64')
general_data['dob_years'] = general_data['dob_years'].replace([0], dob_years_median_int)
# Resuelve los problemas en la columna `dob_years`, si existen


In [30]:
general_data[general_data['dob_years'] == 0].count()
# Comprueba el resultado - asegúrate de que esté arreglado


children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

No se logra visualizar datos que puedan causar problemas en la columna "family status", por lo que no se usara ningun tipo de accion.

In [31]:
# Veamos los valores de la columna
general_data['family_status'].apply(type).unique()

array([<class 'str'>], dtype=object)

In [32]:
# Aborda los valores problemáticos en `family_status`, si existen



In [33]:
# Comprueba el resultado - asegúrate de que esté arreglado


Hay un valor "XNA" en la columna "gender", por lo que al ver que es solo un valor, se eliminara esa columna ya que no tendra mucha relevancia con nuestro analisis y puede malograrlo si se deja intacto.

In [34]:
# Veamos los valores en la columna
general_data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [35]:
general_data[general_data['gender'] == 'XNA']

children  days_employed  dob_years     education  education_id  \
10701         0      98.275021         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  
10701     32624.825  buy real estate

In [36]:
# Aborda los valores problemáticos, si existen
general_data.drop(10701, axis=0, inplace=True)

In [37]:
# Comprueba el resultado - asegúrate de que esté arreglado
general_data['gender'].unique()


array(['F', 'M'], dtype=object)

No se logra visualizar datos que puedan causar problemas en la columna "income_type", por lo que no se usara ningun tipo de accion.

In [38]:
# Veamos los valores en la columna
general_data['income_type'].apply(type).unique()

array([<class 'str'>], dtype=object)

In [39]:
# Aborda los valores problemáticos, si existen

In [40]:
# Comprueba el resultado - asegúrate de que esté arreglado



<div class="alert alert-success">
<h2> Comentarios del revisor V1</h2>

Muy buena decisión eliminando ese género inconsistente y reemplazando el valor de 0 en `dob_years`.

</div>

Lo que podria considerar duplicados seria el proposito para el prestamo, es decir, la columna "purpose". Para solucionarlo voy a reemplazar los valores unicos que son relacionados un solo valor que los represente.

In [41]:
# Comprobar los duplicados
general_data['purpose'].describe()

count                21524
unique                  38
top       wedding ceremony
freq                   797
Name: purpose, dtype: object

In [42]:
# Aborda los duplicados, si existen
for x in general_data['purpose']:
    if 'car' in x:
        general_data['purpose'] = general_data['purpose'].replace(['car purchase', 'buying a second-hand car', 'buying my own car', 'cars', 'to own a car', 'purchase of a car', 'to buy a car' ], 'car')
    elif 'real estate' in x:
        general_data['purpose'] = general_data['purpose'].replace(['buy real estate', 'buy residential real estate', 'buy commercial real estate', 'transactions with commercial real estate', 'building a real estate', 'transactions with my real estate', 'real estate transactions'], 'real estate')
    elif 'education' in x:
        general_data['purpose'] = general_data['purpose'].replace(['supplementary education', 'education', 'getting an education', 'to get a supplementary education', 'getting higher education', 'profile education', 'university education'], 'education motives')
    elif 'educated' in x:
        general_data['purpose'] = general_data['purpose'].replace(['to become educated'], 'education motives')
    elif 'housing' in x:
        general_data['purpose'] = general_data['purpose'].replace(['housing transactions', 'housing', 'housing renovation'], 'housing')
    elif 'wedding' in x:
        general_data['purpose'] = general_data['purpose'].replace(['to have a wedding', 'having a wedding', 'wedding ceremony'], 'wedding')
    elif 'house' in x:
        general_data['purpose'] = general_data['purpose'].replace(['purchase of the house', 'purchase of the house for my family', 'purchase of my own house'], 'house purchase')
    elif 'property' in x:
        general_data['purpose'] = general_data['purpose'].replace(['construction of own property', 'property', 'building a property', 'buying property for renting out'], 'property')


In [43]:
# Última comprobación para ver si tenemos duplicados
general_data['purpose'].describe()

count           21524
unique              9
top       real estate
freq             4477
Name: purpose, dtype: object

In [44]:
# Comprueba el tamaño del conjunto de datos que tienes ahora, después de haber ejecutado estas primeras manipulaciones
general_data.size

258288

Ha cambiado el dataframe, pues datos problematicos que puedan afectar nuestro analisis han sido eliminados o sustituidos por otros, especificamente se ha eliminado una fila o entrada del dataframe.

# Trabajar con valores ausentes

Se puede agrupar el tipo de educacion y el estado civil por valores que estan representados en la misma tabla, que son sus respectivos IDs.

In [45]:
# Encuentra los diccionarios
print(general_data.set_index('education')['education_id'].to_dict())
print(general_data.set_index('family_status')['family_status_id'].to_dict())


{"bachelor's degree": 0, 'secondary education': 1, 'some college': 2, 'primary education': 3, 'graduate degree': 4}
{'married': 0, 'civil partnership': 1, 'widow / widower': 2, 'divorced': 3, 'unmarried': 4}


### Restaurar valores ausentes en `total_income`

Voy a crear una categoria de edad que los separa en grupos desde un adulto joven a un adulto mayor. Esto nos ayudara para poder clasificarlos y en caso sea necesario usar esta clasificacion para restaurar los valores ausentes ya sea en total_income o days_employed


In [46]:
# Vamos a escribir una función que calcule la categoría de edad
def age_group(x):
    if 0 <= x <=30:
        return 'young adult'
    elif 30 <= x <= 59:
        return 'mid adult'
    elif 60 <= x <= 90:
        return 'third age'

In [47]:
# Prueba si la función funciona bien
age_group(50)

'mid adult'

In [48]:
# Crear una nueva columna basada en la función
general_data['age_group'] = general_data['dob_years'].apply(age_group)

In [49]:
# Comprobar cómo los valores en la nueva columna
general_data.groupby('age_group')[['dob_years']].count()


dob_years
age_group             
mid adult        15285
third age         2517
young adult       3722

In [50]:
# Crea una tabla sin valores ausentes y muestra algunas de sus filas para asegurarte de que se ve bien
second_table = general_data[general_data['days_employed'].notnull()]
second_table.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19350 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          19350 non-null  int64  
 1   days_employed     19350 non-null  float64
 2   dob_years         19350 non-null  int64  
 3   education         19350 non-null  object 
 4   education_id      19350 non-null  int64  
 5   family_status     19350 non-null  object 
 6   family_status_id  19350 non-null  int64  
 7   gender            19350 non-null  object 
 8   income_type       19350 non-null  object 
 9   debt              19350 non-null  int64  
 10  total_income      19350 non-null  float64
 11  purpose           19350 non-null  object 
 12  age_group         19350 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


In [51]:
# Examina los valores medios de los ingresos en función de los factores que identificaste
general_data.groupby(['education_id', 'family_status_id'])[['total_income']].mean().reset_index()

education_id  family_status_id  total_income
0              0                 0  33719.993035
1              0                 1  32982.368398
2              0                 2  28042.941435
3              0                 3  32963.499089
4              0                 4  32066.666118
5              1                 0  24706.166301
6              1                 1  24739.490095
7              1                 2  22260.731241
8              1                 3  24876.901784
9              1                 4  24685.396198
10             2                 0  29849.812029
11             2                 1  27984.916629
12             2                 2  24420.224846
13             2                 3  34811.298308
14             2                 4  27985.427769
15             3                 0  22489.557931
16             3                 1  20026.139920
17             3                 2  18001.459679
18             3                 3  19433.435750
19             3                 4  21208.299563
20             4                 0  29359.245250
21             4                 3  18551.846000
22             4                 4  31771.321000

In [52]:
# Examina los valores medianos de los ingresos en función de los factores que identificaste
general_data.groupby(['education_id', 'family_status_id'])[['total_income']].median().reset_index()

education_id  family_status_id  total_income
0              0                 0    28466.2870
1              0                 1    28353.3245
2              0                 2    23789.3965
3              0                 3    27632.7540
4              0                 4    27262.2795
5              1                 0    21987.1500
6              1                 1    21847.7850
7              1                 2    20363.1200
8              1                 3    21914.2800
9              1                 4    21731.8650
10             2                 0    26113.4020
11             2                 1    25709.8980
12             2                 2    20531.8480
13             2                 3    31349.2795
14             2                 4    24563.6940
15             3                 0    20012.3570
16             3                 1    16885.4175
17             3                 2    16632.7265
18             3                 3    18401.2060
19             3                 4    18868.4855
20             4                 0    29345.3940
21             4                 3    18551.8460
22             4                 4    31771.3210

Pienso que la columna que tiene mas relacion es el tipo de trabajo que tienen, pues tiene mucho que ver con los dias trabajados y tipo de ingreso, ya que depende mucho de si una persona esta retirada, etc.


In [53]:
#  Escribe una función que usaremos para completar los valores ausentes
median_results_total_income = general_data.groupby('income_type')['total_income'].median()
def total_income_mean_income_type(row):
    row_income_type = row['income_type']
    row_total_income = row['total_income']
    
    if pd.isna(row_total_income):
        return median_results_total_income[row_income_type]
    else:
        return row_total_income

In [54]:
# Comprueba si funciona
total_income_mean_income_type(general_data.loc[12])

18962.318

In [55]:
# Aplícalo a cada fila
general_data['total_income_clean'] = general_data.apply(total_income_mean_income_type, axis=1)

In [56]:
# Comprueba si tenemos algún error
general_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   children            21524 non-null  int64  
 1   days_employed       19350 non-null  float64
 2   dob_years           21524 non-null  int64  
 3   education           21524 non-null  object 
 4   education_id        21524 non-null  int64  
 5   family_status       21524 non-null  object 
 6   family_status_id    21524 non-null  int64  
 7   gender              21524 non-null  object 
 8   income_type         21524 non-null  object 
 9   debt                21524 non-null  int64  
 10  total_income        19350 non-null  float64
 11  purpose             21524 non-null  object 
 12  age_group           21524 non-null  object 
 13  total_income_clean  21524 non-null  float64
dtypes: float64(3), int64(5), object(6)
memory usage: 3.0+ MB


Por ahora no se ha encontrado ningun error, por lo que no se ve necesario hacer el siguiente paso.


In [57]:
# Reemplazar los valores ausentes si hay algún error


In [58]:
# Comprobar el número de entradas en las columnas
total_income_clean_count = general_data['total_income_clean'].count()
children_count = general_data['children'].count()

print(f'Las entradas totales en total_income_clean_count son: {total_income_clean_count}, y en children: {children_count}, por lo tanto es acertado decir que fue exitosa la funcion.')

Las entradas totales en total_income_clean_count son: 21524, y en children: 21524, por lo tanto es acertado decir que fue exitosa la funcion.


<div class="alert alert-success">
<h2> Comentarios del revisor V1</h2>

Muy buen trabajo!
</div>

###  Restaurar valores en `days_employed`

In [59]:
# Distribución de las medianas de `days_employed` en función de los parámetros identificados
median_results_days_employed = general_data.groupby('income_type')['days_employed'].median()

In [60]:
# Distribución de las medias de `days_employed` en función de los parámetros identificados
mean_results_days_employed = general_data.groupby('income_type')['days_employed'].mean()

Voy a utilizar el promedio, ya que seria lo apropiado, pues es una medida cuantitativa.

In [61]:
# Escribamos una función que calcule medias o medianas (dependiendo de tu decisión) según el parámetro identificado
def days_employed_mean_income_type(row):
    row_income_type = row['income_type']
    row_days_employed = row['days_employed']
    
    if pd.isna(row_days_employed):
        return mean_results_days_employed[row_income_type]
    else:
        return row_days_employed

In [62]:
# Comprueba que la función funciona
days_employed_mean_income_type(general_data.loc[12])

15208.478801869218

In [63]:
# Aplicar la función al income_type
# Ya fue aplicado arriba

In [64]:
# Comprueba si la función funcionó
days_employed_mean_income_type(general_data.loc[12])

15208.478801869218

<div class="alert alert-danger">
<h2> Comentarios del revisor</h2>

Creo que hay código duplicado que genera un error, eliminando la última función debería arreglarse.
</div>

<div class="alert alert-info">
<h2> Comentarios del estudiante</h2>
Si, habia un codigo duplicado, esta corregido ahora.
</div>

In [65]:
# Reemplazar valores ausentes
general_data['days_employed_clean'] = general_data.apply(days_employed_mean_income_type, axis=1)
general_data.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   children             21524 non-null  int64  
 1   days_employed        19350 non-null  float64
 2   dob_years            21524 non-null  int64  
 3   education            21524 non-null  object 
 4   education_id         21524 non-null  int64  
 5   family_status        21524 non-null  object 
 6   family_status_id     21524 non-null  int64  
 7   gender               21524 non-null  object 
 8   income_type          21524 non-null  object 
 9   debt                 21524 non-null  int64  
 10  total_income         19350 non-null  float64
 11  purpose              21524 non-null  object 
 12  age_group            21524 non-null  object 
 13  total_income_clean   21524 non-null  float64
 14  days_employed_clean  21524 non-null  float64
dtypes: float64(4), int64(5), object(6)
m

In [66]:
# Comprueba las entradas en todas las columnas: asegúrate de que hayamos corregido todos los valores ausentes
days_employed_clean_count = general_data['days_employed_clean'].count()
children_count = general_data['children'].count()

print(f'Las entradas totales en days_employed_clean_count son: {days_employed_clean_count}, y en children: {children_count}, por lo tanto es acertado decir que fue exitosa la funcion.')

Las entradas totales en days_employed_clean_count son: 21524, y en children: 21524, por lo tanto es acertado decir que fue exitosa la funcion.


In [67]:
# Muestra los valores de los datos seleccionados para la clasificación
general_data['purpose'].describe()


count           21524
unique              9
top       real estate
freq             4477
Name: purpose, dtype: object

Vamos a comprobar los valores únicos

In [68]:
# Comprobar los valores únicos
general_data['purpose'].unique()

array(['house purchase', 'car', 'education motives', 'wedding', 'housing',
       'real estate', 'property', 'second-hand car purchase',
       'going to university'], dtype=object)

Como antes se reemplazo los valores de purpose por temas en comun, no es absolutamente necesario agruparlos en categorias, ya que no hay valores similares, pero por el proposito del proyecto de todas maneras se lo puede realizar.


In [69]:
# Escribamos una función para clasificar los datos en función de temas comunes
def purpose_group(x):
    if 'car' in x:
        return 'car'
    elif 'real estate' in x:
        return 'real estate motives'
    elif 'education' in x:
        return 'education motives'
    elif 'educated' in x:
        return 'education motives'
    elif 'housing' in x:
        return 'house purposes'
    elif 'wedding' in x:
        return 'wedding fonds'
    elif 'house' in x:
        return 'house purchase'
    elif 'property' in x:
        return 'house purposes'
    elif 'university' in x:
        return 'education motives'

In [70]:
# Crea una columna con las categorías y cuenta los valores en ellas

general_data['purpose_group'] = general_data['purpose'].apply(purpose_group)
general_data['purpose_group'].value_counts()

real estate motives    4477
house purposes         4454
car                    4315
education motives      4022
wedding fonds          2348
house purchase         1908
Name: purpose_group, dtype: int64

In [71]:
# Revisar todos los datos numéricos en la columna seleccionada para la clasificación
general_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   children             21524 non-null  int64  
 1   days_employed        19350 non-null  float64
 2   dob_years            21524 non-null  int64  
 3   education            21524 non-null  object 
 4   education_id         21524 non-null  int64  
 5   family_status        21524 non-null  object 
 6   family_status_id     21524 non-null  int64  
 7   gender               21524 non-null  object 
 8   income_type          21524 non-null  object 
 9   debt                 21524 non-null  int64  
 10  total_income         19350 non-null  float64
 11  purpose              21524 non-null  object 
 12  age_group            21524 non-null  object 
 13  total_income_clean   21524 non-null  float64
 14  days_employed_clean  21524 non-null  float64
 15  purpose_group        21524 non-null 

In [72]:
# Obtener estadísticas resumidas para la columna
general_data['total_income_clean'].describe()


count     21524.000000
mean      26435.618312
std       15687.283573
min        3306.762000
25%       17247.356500
50%       22815.103500
75%       31287.232000
max      362496.645000
Name: total_income_clean, dtype: float64

Para la cantidad de hijos voy a usar de 1 a 3 como la cantidad promedio(average), mientras que para salarios voy a usar los valores de 0 al 25% para un salario bajo, de 25% al 50% para un salario promedio, y finalmente de 50% para arriba como un salario alto.

In [73]:
# Crear una función para clasificar en diferentes grupos numéricos basándose en rangos
def children_group(x):
    if x == 0:
        return 'no kids'
    if 0 < x <= 3:
        return 'average'
    if 3 <= x <= 5:
        return 'high'

def salary_group(x):
    if 0 <= x <= 17247.356500:
        return 'low salary'
    if 17247.356500 <= x <= 22815.103500:
        return 'average salary'
    if 22815.103500 <= x:
        return 'high salary'

In [74]:
# Crear una columna con categorías
general_data['children_group'] = general_data['children'].apply(children_group)
general_data['salary_group'] = general_data['total_income_clean'].apply(salary_group)

In [75]:
# Contar los valores de cada categoría para ver la distribución
print(general_data['children_group'].value_counts())
print(general_data['salary_group'].value_counts())

no kids    14271
average     7203
high          50
Name: children_group, dtype: int64
high salary       10619
average salary     5524
low salary         5381
Name: salary_group, dtype: int64


## Comprobación de las hipótesis


**¿Existe una correlación entre tener hijos y pagar a tiempo?**

In [76]:
children_group_1 = (general_data[(general_data['children_group'] == 'no kids') & (general_data['debt'] == 0)].count())
children_group_2 = (general_data[(general_data['children_group'] == 'average') & (general_data['debt'] == 0)].count())
children_group_3 = (general_data[(general_data['children_group'] == 'high') & (general_data['debt'] == 0)].count())
children_group_1_debt = (general_data[(general_data['children_group'] == 'no kids') & (general_data['debt'] == 1)].count())
children_group_2_debt = (general_data[(general_data['children_group'] == 'average') & (general_data['debt'] == 1)].count())
children_group_3_debt = (general_data[(general_data['children_group'] == 'high') & (general_data['debt'] == 1)].count())

In [77]:
# Comprueba los datos sobre los hijos y los pagos puntuales
children_group_1_total = 13199
children_group_2_total = 6538
children_group_3_total = 46
children_group_1_debt_total = 1072
children_group_2_debt_total = 665
children_group_3_debt_total = 4
# Calcular la tasa de incumplimiento en función del número de hijos
print(f'La tasa de incumplimiento cuando tienen 0 hijos es: {(children_group_1_debt_total/children_group_1_total):%}.')
print(f'La tasa de incumplimiento cuando tienen de 1 a 3 hijos es: {(children_group_2_debt_total/children_group_2_total):%}.')
print(f'La tasa de incumplimiento cuando tienen de 3 a 5 hijos es: {(children_group_3_debt_total/children_group_3_total):%}.')
grupo_1 = (children_group_2_debt_total/children_group_2_total) * 100
grupo_2 = (children_group_3_debt_total/children_group_3_total) * 100
total_deuda_hijos = (grupo_1+grupo_2)
print(f'La tasa de incumplimiento cuando tienen hijos y cuando no es: {(total_deuda_hijos-((children_group_1_debt_total/children_group_1_total)) * 100)}.')

La tasa de incumplimiento cuando tienen 0 hijos es: 8.121827%.
La tasa de incumplimiento cuando tienen de 1 a 3 hijos es: 10.171306%.
La tasa de incumplimiento cuando tienen de 3 a 5 hijos es: 8.695652%.
La tasa de incumplimiento cuando tienen hijos y cuando no es: 10.74513097259564.


**Conclusión**

Las personas que tienen de 1 a 3 hijos son las personas que menos cumplen con sus deudas, mientras las que mas cumplen son las personas que no tienen hijos.

**¿Existe una correlación entre la situación familiar y el pago a tiempo?**

In [78]:
family_status_0 = (general_data[(general_data['family_status_id'] == 0) & (general_data['debt'] == 0)].count())
family_status_1 = (general_data[(general_data['family_status_id'] == 1) & (general_data['debt'] == 0)].count())
family_status_2 = (general_data[(general_data['family_status_id'] == 2) & (general_data['debt'] == 0)].count())
family_status_3 = (general_data[(general_data['family_status_id'] == 3) & (general_data['debt'] == 0)].count())
family_status_4 = (general_data[(general_data['family_status_id'] == 4) & (general_data['debt'] == 0)].count())
family_status_0_debt = (general_data[(general_data['family_status_id'] == 0) & (general_data['debt'] == 1)].count())
family_status_1_debt = (general_data[(general_data['family_status_id'] == 1) & (general_data['debt'] == 1)].count())
family_status_2_debt = (general_data[(general_data['family_status_id'] == 2) & (general_data['debt'] == 1)].count())
family_status_3_debt = (general_data[(general_data['family_status_id'] == 3) & (general_data['debt'] == 1)].count())
family_status_4_debt = (general_data[(general_data['family_status_id'] == 4) & (general_data['debt'] == 1)].count())

In [79]:
# Comprueba los datos del estado familiar y los pagos a tiempo
family_status_0_total = 11449
family_status_1_total = 3788
family_status_2_total = 897
family_status_3_total = 1110
family_status_4_total = 2539
family_status_0_debt_total = 931
family_status_1_debt_total = 388
family_status_2_debt_total = 63
family_status_3_debt_total = 85
family_status_4_debt_total = 274
# Calcular la tasa de incumplimiento basada en el estado familiar
print(f'La tasa de incumplimiento cuando son casados es: {(family_status_0_debt_total/family_status_0_total):%}.')
print(f'La tasa de incumplimiento cuando son unidos es: {(family_status_1_debt_total/family_status_1_total):%}.')
print(f'La tasa de incumplimiento cuando son viudos/viudas es: {(family_status_2_debt_total/family_status_2_total):%}.')
print(f'La tasa de incumplimiento cuando son divorciados es: {(family_status_3_debt_total/family_status_3_total):%}.')
print(f'La tasa de incumplimiento cuando son solteros es: {(family_status_4_debt_total/family_status_4_total):%}.')

La tasa de incumplimiento cuando son casados es: 8.131715%.
La tasa de incumplimiento cuando son unidos es: 10.242872%.
La tasa de incumplimiento cuando son viudos/viudas es: 7.023411%.
La tasa de incumplimiento cuando son divorciados es: 7.657658%.
La tasa de incumplimiento cuando son solteros es: 10.791650%.


**Conclusión**

Las personas solteras son las que menos tienen responsabilidad con sus deudas, mientras que los mas responsables con sus deudad son los viudos/viudas.

**¿Existe una correlación entre el nivel de ingresos y el pago a tiempo?**

In [80]:
salary_group_1 = (general_data[(general_data['salary_group'] == 'low salary') & (general_data['debt'] == 0)].count())
salary_group_2 = (general_data[(general_data['salary_group'] == 'average salary') & (general_data['debt'] == 0)].count())
salary_group_3 = (general_data[(general_data['salary_group'] == 'high salary') & (general_data['debt'] == 0)].count())
salary_group_debt_1 = (general_data[(general_data['salary_group'] == 'low salary') & (general_data['debt'] == 1)].count())
salary_group_debt_2 = (general_data[(general_data['salary_group'] == 'average salary') & (general_data['debt'] == 1)].count())
salary_group_debt_3 = (general_data[(general_data['salary_group'] == 'high salary') & (general_data['debt'] == 1)].count())

In [81]:
# Comprueba los datos del nivel de ingresos y los pagos a tiempo
salary_1_total = 4954
salary_2_total = 5041
salary_3_total = 9788
salary_1_debt_total = 427
salary_2_debt_total = 483
salary_3_debt_total = 831
# Calcular la tasa de incumplimiento basada en el nivel de ingresos
print(f'La tasa de incumplimiento cuando tienen un salario bajo es: {(salary_1_debt_total/salary_1_total):%}.')
print(f'La tasa de incumplimiento cuando tienen un salario promedio es: {(salary_2_debt_total/salary_2_total):%}.')
print(f'La tasa de incumplimiento cuando tienen un salario alto es: {(salary_3_debt_total/salary_3_total):%}.')

La tasa de incumplimiento cuando tienen un salario bajo es: 8.619298%.
La tasa de incumplimiento cuando tienen un salario promedio es: 9.581432%.
La tasa de incumplimiento cuando tienen un salario alto es: 8.489988%.


**Conclusión**

Las personas con un salario promedio son las que tienden a no cumplir con sus deudas, mientras que los que mas cumplen con esta obligacion son los de salario alto.

**¿Cómo afecta el propósito del crédito a la tasa de incumplimiento?**

In [82]:
house_purchase = (general_data[(general_data['purpose_group'] == 'house purchase') & (general_data['debt'] == 0)].count())
car = (general_data[(general_data['purpose_group'] == 'car') & (general_data['debt'] == 0)].count())
education_motives = (general_data[(general_data['purpose_group'] == 'education motives') & (general_data['debt'] == 0)].count())
wedding = (general_data[(general_data['purpose_group'] == 'wedding fonds') & (general_data['debt'] == 0)].count())
house_purposes = (general_data[(general_data['purpose_group'] == 'house purposes') & (general_data['debt'] == 0)].count())
real_estate_motives = (general_data[(general_data['purpose_group'] == 'real estate motives') & (general_data['debt'] == 0)].count())
house_purchase_debt = (general_data[(general_data['purpose_group'] == 'house purchase') & (general_data['debt'] == 1)].count())
car_debt = (general_data[(general_data['purpose_group'] == 'car') & (general_data['debt'] == 1)].count())
education_motives_debt = (general_data[(general_data['purpose_group'] == 'education motives') & (general_data['debt'] == 1)].count())
wedding_debt = (general_data[(general_data['purpose_group'] == 'wedding fonds') & (general_data['debt'] == 1)].count())
house_purposes_debt = (general_data[(general_data['purpose_group'] == 'house purposes') & (general_data['debt'] == 1)].count())
real_estate_motives_debt = (general_data[(general_data['purpose_group'] == 'real estate motives') & (general_data['debt'] == 1)].count())

In [83]:
# Consulta los porcentajes de tasa de incumplimiento para cada propósito del crédito y analízalos
house_purchase_total = 1781
car_total = 3912
education_motives_total = 3652
wedding_total = 2162
house_purposes_total = 4135
real_estate_motives_total = 4141
house_purchase_debt_total = 127
car_debt_total = 403
education_motives_debt_total = 370
wedding_debt_total = 186
house_purposes_debt_total = 319
real_estate_motives_debt_total = 336
print(f'La tasa de incumplimiento cuando quieren comprarse una casa es: {(house_purchase_debt_total/house_purchase_total):%}.')
print(f'La tasa de incumplimiento cuando quieren comprarse un carro es: {(car_debt_total/car_total):%}.')
print(f'La tasa de incumplimiento cuando quieren invertir en educacion es: {(education_motives_debt_total/education_motives_total):%}.')
print(f'La tasa de incumplimiento cuando quieren invertir en fondos para boda es: {(wedding_debt_total/wedding_total):%}.')
print(f'La tasa de incumplimiento cuando quieren invertir en propiedades es: {(house_purposes_debt_total/house_purposes_total):%}.')
print(f'La tasa de incumplimiento cuando quieren invertir en bienes raices es: {(real_estate_motives_debt_total/real_estate_motives_total):%}.')

La tasa de incumplimiento cuando quieren comprarse una casa es: 7.130825%.
La tasa de incumplimiento cuando quieren comprarse un carro es: 10.301636%.
La tasa de incumplimiento cuando quieren invertir en educacion es: 10.131435%.
La tasa de incumplimiento cuando quieren invertir en fondos para boda es: 8.603145%.
La tasa de incumplimiento cuando quieren invertir en propiedades es: 7.714631%.
La tasa de incumplimiento cuando quieren invertir en bienes raices es: 8.113982%.


<div class="alert alert-success">
<h2> Comentarios del revisor V1</h2>

Felicitaciones, muy buen trabajo en general!
</div>

<div class="alert alert-success">
<h2> Comentarios del revisor V2</h2>

Excelente trabajo, tienes todo muy claro!
</div>

**Conclusión**

Las personas que quieren comprarse un carro en este caso son los que menos cumplen con su deuda, mientras que los que mas cumplen son los que quieren comprarse una casa.


# Conclusión general 


# 1.- La correlacion entre tener hijos y tener deudas, a no tener hijos y tener deudas existe y fue 10.74513097259564%.

# 2.- La correlacion entre el estado civil y estar endeudados existe y fue allado previamente.

# 3.- La correlacion entre el tipo de salario que gana cada persona y tener deudas existe, siendo los que tienen salario promedio los que menos cumplen con sus deudas, y los que tienen mejor salario cumplen mas con sus deudas.

# 4 .- El proposito de prestamo tambien afecta al si tienen deuda o no, pues los que quieren comprarse carros tienen menos tendencia a cumplir con su obligacion, mientras que los que quieren comprarse una casa cumplen mas con esta.

# Conclusion .- La conclusion general es que dependiendo de muchos factores es posible examinar si un candidato es apto para el prestamo, pues depende de sus ingresos, cantidad de hijos, estado civil, entre otros. 